src/batchnorm.jl

```diff
@@ -0,0 +1,485 @@
+"""
+
+`BNMoments` is a high-level data structure used to store running mean and running variance
+of batch normalization.
+
+# Fields
+ `momentum::AbstractFloat`: A real number between 0 and 1 to be used as the scale of
+  last mean and variance. The existing running mean or variance is multiplied with 
+  (1-momentum).
+ 
+ `mean`: The running mean.
+
+ `var`: The running variance.
+ 
+ `meaninit`: The function used for initialize the running mean. Should either be `nothing` or
+of the form `(eltype, dims...) -> data`. `zeros` is a good option.
+
+ `varinit`: The function used for initialize the running variance. Should either be `nothing` or
+`(eltype, dims...) -> data`. `ones` is a good option.
+
+# Constructors
+ `BNMoments(;momentum=0.1, meaninit=zeros, varinit=ones)` sets `mean` and `var` to nothing.
+When this constructor is used, `batchnorm` functions will initialize `mean` and `var` during
+the first training iteration to have relevant dimensions and match the data type of the input.
+
+ `BNMoments(momentum, mean, var)` can be used directly load moments from data. `meaninit` and
+`varinit` are made nothing and they won't be called by `batchnorm`. Its user's responsibility
+ to initialize mean and var with right dimensions and types.
+
+"""
+type BNMoments
+    momentum::AbstractFloat
+    mean
+    var
+    meaninit
+    varinit
+end
+
+BNMoments(;momentum=0.1, meaninit=zeros, varinit=ones, o...) =
+    BNMoments(momentum, nothing, nothing, meaninit, varinit)
+
+BNMoments(momentum, mean, var) = BNMoments(momentum, mean, var, nothing, nothing)
+
+
+#TODO: improve the documentation. Explain what batchnorm is.
+"""
+
+`batchnorm(x[, moments, param]; kwargs...)` performs batch normalization to `x`
+with optional scaling factor and bias stored in `w`. Operation performed is spatial 
+batch normalization if x is 4d or 5d. `moments` is a `BNMoments` object that stores 
```

Comment:
Are there other options than 4d and 5d?  What is spatial batch normalization, are there other alternatives?

Reply:
In CNNs, it is assumed each channel is a single feature, so batch norm is performed by averaging the input in all dimensions other than the channel dimesion (dimension `ndims(x)-1`).  There is also a per-activation batch normalization, which averages all dimensions other than the batch dimension. The two modes are identical when the input is 2d. I didn't support per activation mode for 4d and 5d since it is not common and user can perform this without any overhead by calling `reshape(batchnorm(mat(x)), size(x))`.

The normalization dimension will be added to the doc. We also support 2d inputs, which I'll  ad to the documentation. I'm not sure what to do for 3d inputs, so I didn't support them for now.